service/Program.cs

```diff
@@ -0,0 +1,81 @@
+﻿using Azure.IoT.ModelsRepository;
+using Microsoft.Azure.Devices;
+using Microsoft.Azure.Devices.Serialization;
+using Microsoft.Azure.DigitalTwins.Parser;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using System.Threading.Tasks;
+using System.Web;
+
+namespace service
+{
+
+    class Program
+    {
+        static readonly string cs = Environment.GetEnvironmentVariable("HUB_CS");
+        static readonly DigitalTwinClient dtc = DigitalTwinClient.CreateFromConnectionString(cs);
+        static readonly string deviceId = "self2";
+
+        static async Task Main(string[] args)
+        {
+            var twinResponse = await dtc.GetDigitalTwinAsync<BasicDigitalTwin>(deviceId);
+            Console.WriteLine($"Device '{deviceId}' announced: {twinResponse.Body.Metadata.ModelId}\n");
+
+            var model = await ResolveAndParse(twinResponse.Body);
+            Console.WriteLine("\nModel Parsed !! \n\n");
+            model.ToList().ForEach(i => Console.WriteLine(i.Key));
+
+            Console.ReadLine();
+        }
+
+        private static async Task<IReadOnlyDictionary<Dtmi, DTEntityInfo>> ResolveAndParse(BasicDigitalTwin twin)
+        {
+            IReadOnlyDictionary<Dtmi, DTEntityInfo> model;
+
+            string repo = "https://raw.githubusercontent.com/iotmodels/iot-plugandplay-models/selfdescribing";
+            ModelsRepositoryClient dmrClient = new ModelsRepositoryClient(new Uri(repo));
+            ModelParser modelParser = new ModelParser()
+            {
+                DtmiResolver = dmrClient.ParserDtmiResolver
+            };
+
+            Uri modelId = new Uri(twin.Metadata.ModelId);
+            if (modelId.AbsolutePath == "azure:common:SelfDescribing;1")
+            {
+                Console.WriteLine("Device is Self Reporting. Querying device for the model . . ");
+
+                var commandResponse = await dtc.InvokeCommandAsync(deviceId, "GetTargetModel");
+                string modelPayload = commandResponse.Body.Payload;
+                Console.Write("Device::GetTargetModel() ok..");
+                string discoveredModelId = JsonDocument.Parse(modelPayload).RootElement.GetProperty("@id").GetString();
+                Console.Write("Discovered ModelId: " + discoveredModelId);
+
+                string targetModelHash = SelfDescribingChecks.GetPropFromModelOrTwin(twin, "tmhash", "targetModelHash");
+                if (!string.IsNullOrEmpty(targetModelHash))
+                {
+
+                    SelfDescribingChecks.CheckHash(targetModelHash, modelPayload);
```

Comment:
My perception of self-describing devices and the hash was that a service app could cache the model along with the hash, and only ask for the model again if the hash changed.

This sample insinuates a service app must confirm the hash, and my take is that is 100% not necessary. IIRC the hash was just one idea of how to derive a unique "version" value, but that it could be anything.

Reply:
Agree it's not 100% necessary, hence the sample only confirms the hash if it's present.

For large scale solutions, like Central, it will require the hash for the next  reasons:
1) To avoid calling each device if 2 or more use the same model
2) To detect changes in the model when the id is the same
3) To enable caching

Reply:
Then let's build the sample to do that. Build a cache, and use that when a device connects to determine whether or not to call `GetModel`.